src/main/java/org/wildfly/security/sasl/util/AuthenticationTimeoutSaslServerFactory.java

```diff
@@ -0,0 +1,128 @@
+ * <p>
+ * This {@link SaslServerFactory} must be outside of the {@link AuthenticationCompleteCallbackSaslServerFactory} in the
+ * chain of SASL server factories.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public final class AuthenticationTimeoutSaslServerFactory extends AbstractDelegatingSaslServerFactory {
+
+    /**
+     * The default amount of time, in seconds, after which an authentication attempt should time out.
+     */
+    public static final long DEFAULT_TIMEOUT = 150;
+
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param delegate the delegate {@code SaslServerFactory}
+     * @param scheduledExecutorService the scheduled executor to use to handle authentication timeout tasks
+     */
+    public AuthenticationTimeoutSaslServerFactory(final SaslServerFactory delegate, final ScheduledExecutorService scheduledExecutorService) {
+        super(delegate);
+        Assert.checkNotNullParam("scheduledExecutorService", scheduledExecutorService);
+        this.scheduledExecutorService = scheduledExecutorService;
+    }
+
+    @Override
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        final long timeout;
+        if (props.containsKey(WildFlySasl.AUTHENTICATION_TIMEOUT)) {
+            timeout = Long.parseLong((String) props.get(WildFlySasl.AUTHENTICATION_TIMEOUT));
+        } else {
+            timeout = DEFAULT_TIMEOUT;
+        }
+
+        final SaslServer delegateSaslServer = delegate.createSaslServer(mechanism, protocol, serverName, props, cbh);
+        return (delegateSaslServer == null) ? null : new DelegatingTimeoutSaslServer(delegateSaslServer, scheduledExecutorService, timeout);
+    }
+
+    private static class DelegatingTimeoutSaslServer extends AbstractDelegatingSaslServer {
+        private boolean complete;
+        private ScheduledFuture<Void> timeoutTask;
+
+        DelegatingTimeoutSaslServer(final SaslServer delegate, final ScheduledExecutorService scheduledExecutorService, final long timeout) {
+            super(delegate);
+
+            // Schedule a task to terminate the authentication attempt if it takes too long
+            timeoutTask = scheduledExecutorService.schedule(() -> {
+                synchronized (this) {
+                    if (! (isComplete() && ! complete)) {
+                        dispose();
+                    } else {
+                        complete = true;
+                    }
+                    return null;
+                }
+            }, timeout, TimeUnit.SECONDS);
+        }
+
+        @Override
+        public synchronized byte[] evaluateResponse(final byte[] response) throws SaslException {
+            try {
```

Comment:
One problem here - if the authentication did in fact time out, this method should be failing without calling the delegate (as it can easily have been called after the timeout occurred).  The delegate will in fact be invalid in this case (as it will have been disposed) thus there might be some undesirable behavior as a result.